src/App/Fossa/Analyze/FossaV1.hs

```diff
@@ -0,0 +1,307 @@
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Req
+import qualified Network.HTTP.Types as HTTP
+import Prologue
+import Srclib.Converter (toSourceUnit)
+import Srclib.Types
+import Data.Maybe (catMaybes)
+
+newtype FossaReq a = FossaReq { unFossaReq :: ErrorC FossaError IO a }
+  deriving (Functor, Applicative, Monad, MonadIO)
+
+instance MonadHttp FossaReq where
+  handleHttpException = FossaReq . throwError . mangleError
+
+fossaReq :: MonadIO m => FossaReq a -> m (Either FossaError a)
+fossaReq = liftIO . runError @FossaError . unFossaReq
+
+-- TODO: git commit?
+cliVersion :: Text
+cliVersion = "spectrometer"
+
+uploadUrl :: Url scheme -> Url scheme
+uploadUrl baseurl = baseurl /: "api" /: "builds" /: "custom"
+
+-- FIXME: we only want to include organizationId for "archive" and "custom"
+-- TODO: need to normalize "git" projects
+-- render a locator for use in fossa API urls
+renderLocatorUrl :: Int -> Locator -> Text
+renderLocatorUrl orgId Locator{..} =
+  locatorFetcher <> "+" <> T.pack (show orgId) <> "/" <> locatorProject <> "$" <> fromMaybe "" locatorRevision
+
+data UploadResponse = UploadResponse
+  { uploadLocator :: Text
+  , uploadError   :: Maybe Text
+  } deriving (Eq, Ord, Show, Generic)
+
+instance FromJSON UploadResponse where
+  parseJSON = withObject "UploadResponse" $ \obj ->
+    UploadResponse <$> obj .: "locator"
+                   <*> obj .:? "error"
+
+data FossaError
+  = InvalidProjectOrRevision HttpException
+  | NoPermission HttpException
+  | JsonDeserializeError String
+  | OtherError HttpException
+  deriving (Show, Generic)
+
+data ProjectRevision = ProjectRevision
+  { projectName :: Text
+  , projectRevision :: Text
+  } deriving (Eq, Ord, Show, Generic)
+
+data ProjectMetadata = ProjectMetadata
+  { projectTitle :: Maybe Text
+  , projectBranch :: Maybe Text
+  , projectUrl :: Maybe Text
+  , projectJiraKey :: Maybe Text
+  , projectLink :: Maybe Text
+  , projectTeam :: Maybe Text
+  , projectPolicy :: Maybe Text
+  } deriving (Eq, Ord, Show, Generic)
+
+uploadAnalysis
+  :: Url 'Https -- ^ base url
+  -> Text -- ^ api key
+  -> ProjectRevision
+  -> ProjectMetadata
+  -> [Project]
+  -> FossaReq UploadResponse
+uploadAnalysis baseurl key ProjectRevision{..} metadata projects = do
+  let filteredProjects = filter (isProductionPath . projectPath) projects
+      sourceUnits = fromMaybe [] $ traverse toSourceUnit filteredProjects
+      opts = "locator" =: renderLocator (Locator "custom" projectName (Just projectRevision))
+          <> "v" =: cliVersion
+          <> "managedBuild" =: True
+          <> "title" =: fromMaybe projectName (projectTitle metadata)
+          <> header "Authorization" ("token " <> encodeUtf8 key)
+          <> mkMetadataOpts metadata
+  resp <- req POST (uploadUrl baseurl) (ReqBodyJson sourceUnits) jsonResponse opts
+  pure (responseBody resp)
+
+mkMetadataOpts :: ProjectMetadata -> Option scheme
+mkMetadataOpts ProjectMetadata{..} = mconcat $ catMaybes $
```

Comment:
this is clever, but also probably the best way to do this(?)

we have a bunch of optional flags in ProjectMetadata that were passed from the command line, and we need to turn them into query options:

some intermediate types here:
`projectBranch :: Maybe Text`
fmapping in the query option constructor gives us:
`("branch" =:) <$> projectBranch :: Maybe (Option scheme)`
we can use catMaybes to filter out the `Nothing`s:
`catMaybes :: [Maybe (Option scheme)] -> [Option scheme]`
and finally we can use mconcat to combine the query param Options with `<>`
`mconcat :: [Option scheme] -> Option scheme`